File: 06-linked_list/00023-Merge_k_Sorted_Lists/00023-merge_k_sorted_lists.cpp
// 23: Merge k Sorted Lists
// https://leetcode.com/problems/merge-k-sorted-lists


#include <iostream>
#include <vector>
#include <queue>

using namespace std;

struct ListNode {
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};

void llInput(ListNode** head, vector<int> inputs) {
    for (int i=inputs.size()-1; i>=0; i--) {
        ListNode* node = new ListNode(inputs[i], *head);
        *head = node;
    }
}

void llOutput(ListNode* head) {
    while (head != NULL) {
        cout << head->val << " -> ";
        head = head->next;
    }
    cout << "NULL" << endl;
}



class Solution {
private:
    struct compare {
        bool operator()(const ListNode* l, const ListNode* r) {
            return l->val > r->val;
        }
    };

public:
    // SOLUTION
    ListNode* mergeKLists(vector<ListNode*> &lists) {
        priority_queue<ListNode*, vector<ListNode*>, compare> heap;
        ListNode head;
        ListNode *c = &head;
        int i, k, n = lists.size();

        for (i=0; i<n; i++)
            if (lists[i]) heap.push(lists[i]);

        while (!heap.empty()) {
            c->next = heap.top();
            heap.pop();
            c = c->next;
            if (c->next) heap.push(c->next);
        }

        return head.next;
    }
};



int main() {
    Solution o;
    ListNode* ll1 = NULL;
    ListNode* ll2 = NULL;
    ListNode* ll3 = NULL;

    // INPUT
    vector<int> li1 = {1,4,5};
    vector<int> li2 = {1,3,4};
    vector<int> li3 = {2,6};
    llInput(&ll1, li1);
    llInput(&ll2, li2);
    llInput(&ll3, li3);
    vector<ListNode*> lists = {ll1, ll2, ll3};

    // OUTPUT
    auto result = o.mergeKLists(lists);
    llOutput(result);

    return 0;
}

File: 06-linked_list/00023-Merge_k_Sorted_Lists/00023-merge_k_sorted_lists.java
// 23: Merge k Sorted Lists
// https://leetcode.com/problems/merge-k-sorted-lists


import java.util.PriorityQueue;

class Solution {
    public static class ListNode {
        int val;
        ListNode next;
        ListNode() {}
        ListNode(int val) { this.val = val; }
        ListNode(int val, ListNode next) { this.val = val; this.next = next; }
    }

    public static class LinkedList {
        ListNode head;
        LinkedList() { this.head = null; }
    }

    public static void llInput(LinkedList list, int[] inputs) {
        for (int i=inputs.length-1; i>=0; i--) {
            ListNode node = new ListNode(inputs[i], list.head);
            list.head = node;
        }
    }

    public static void llOutput(ListNode head) {
        while (head != null) {
            System.out.print(head.val + " -> ");
            head = head.next;
        }
        System.out.println("NULL");
    }


    // SOLUTION
    public ListNode mergeKLists(ListNode[] lists) {
        PriorityQueue<ListNode> heap = new PriorityQueue<>(lists.length, (l,r)-> l.val - r.val);
        ListNode head = new ListNode();
        ListNode c = head;

        for (ListNode node : lists)
            if (node!=null)
                heap.add(node);

        while (!heap.isEmpty()) {
            c.next = heap.poll();
            c = c.next;
            if (c.next!=null) heap.add(c.next);
        }

        return head.next;
    }


    public static void main(String[] args) {
        Solution o = new Solution();
        LinkedList ll1 = new LinkedList();
        LinkedList ll2 = new LinkedList();
        LinkedList ll3 = new LinkedList();

        // INPUT
        int[] li1 = {1,4,5};
        int[] li2 = {1,3,4};
        int[] li3 = {2,6};
        llInput(ll1, li1);
        llInput(ll2, li2);
        llInput(ll3, li3);
        ListNode[] lists = {ll1.head, ll2.head, ll3.head};

        // OUTPUT
        var result = o.mergeKLists(lists);
        llOutput(result);
    }
}